### DISCUSSION TITLE
LDM Notes for July 31st, 2023

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#7408] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-07-31.md

## Agenda

- Primary constructor parameters and `readonly`

**Comment by qrli** [comment]
> We also investigated whether we can confidently promote primary constructor parameters to full type members in a simple fashion: could we, for example, have `private` automatically turn into a `private readonly` field, and `public` automatically turn into a `public { get; init; }` property?

My understanding of current design is that primary constructor in non-record types won't generate properties. Will it or this paragraph is talking about possible future extensions?

> In terms of evolution, there also a number of LDM members who are concerned about the slippery slope that is evolution in the first place. ... At what point do we say that we've gone far enough and that any further member-like things need to be actual type members?

If this is talking about bringing in many of record features, either now of potentially in future, I can understand. But if this is talking about primary constructor, I'm confused as it does not sound relevant.

> we think that there's a core set of rules around mutation that apply everywhere in the language, except for records. records exist to allow C# users to work with data more easily, and they contain a number of opinions that make this easier, from value-based equality to readonlyness for reference types. We think that this is the C# we want to explain; the alternative would be explaining that primary constructors differ from the rest of C# by bringing in opinions from record types. 

My impression is that all your arguments are around on the idea that "primary constructor parameters are normal parameters like in any other functions". If the scope of primary constructor parameters was just within the primary constructor body, as other functions, I'd fully agree. However, _it is visible and mutable in the whole class_, including other files with `partial`. That is _the_ problem, and why we want it to be readonly. Unfortunately, it sounds like this is not even part of your consideration.

**Reply by CyrusNajmabadi** [reply]
>  understanding of current design is that primary constructor in non-record types won't generate properties. 

They will not generate properties with the basic `Type name` syntax for a parameter. 

> or this paragraph is talking about possible future extensions?

It's talking about the future. Where you might be able to do something like `private Type name` for the parameter and the lang will synthesize a matching private field.

**Reply by CyrusNajmabadi** [reply]
> If this is talking about bringing in many of record features, either now of potentially in future, I can understand. But if this is talking about primary constructor, I'm confused as it does not sound relevant.

It's relevant as we are having active discussions about potential future design extensions to primary constructors and how they would work.  

So, when deciding what was best to do here, there were deep discussions about how choices here might play well with potential future work, it how it might play poorly.  Different ldm members had different thoughts on that.

**Reply by CyrusNajmabadi** [reply]
> Unfortunately, it sounds like this is not even part of your consideration.

This was very much discussed.  That doesn't mean it was agreed with though. 

> That is the problem, and why we want it to be readonly

Understood. No one in the team thinks the desire for the parameters to be readonly is strange. Where the debate came down to was about how we felt these should behave *by default*. 

We concluded that "mutable by default" was appropriate and that adding an explicit `readonly ` modifier was the best way to go for the short *and* long term.

**Reply by qrli** [reply]
Thanks for the clarification. That's very nice of you.

I agree with "mutable by default". My concern is really about whether primary constructor parameter is worth to be a special case, like the `foreach` variable. Maybe you did have that discussion and have arguments against it, though not captured in the notes.

**Reply by CyrusNajmabadi** [reply]
> My concern is really about whether primary constructor parameter is worth to be a special case, like the foreach variable

Yes, that's what the entire choice was between.  Either special casing them, or not.  :)

We decided for 'not' :)

**Reply by DavidArno** [reply]
@qrli 
> I agree with "mutable by default"

I don't even remotely agree with the "mutable by default" line. Consider the scoping of out vars and patterns. The "scoped to current block by default" line was abandoned there because changing the rules for these two features brought benefits that were felt to outweigh maintaining consistency. The language team have therefore previously demonstrated a willingness to favour expediency over consistency when they find that inconsistency to be compelling.

We have a situation where there is significant demand for a simple way to make the resultant field from a "captured" PC parameter be readonly. [Back in 2018, @gafter stated](https://github.com/dotnet/csharplang/issues/1278#issuecomment-360619835):

> Recently, we have been working on a number of language changes to make it easier to program with immutable data. Part of the strategy is to make new shorter shortcuts for readonly properties. If the syntax for a read-write property is shorter than the syntax for a read-only property, then we have failed, and the changes would not encourage programming with immutable data as we intended.

Readonly by default for non-record class PC parameters fits that strategy. Coupled with the fact that the scope of PC parameters is different to normal constructor parameters, this clearly offers good enough grounds for the team to adopt that previous used expediency over consistency approach. The fact that the team haven't just done this and instead are promoting the "mutable by default" line suggests there has been an ideology shift within the language team away from favouring immutability. This is really disappointing.

**Reply by CyrusNajmabadi** [reply]
> line was abandoned there because changing the rules for these two features brought benefits that were felt to outweigh maintaining consistency. 

Note: scoping to the current block *is* consistent with the rest of the language. The original designs for these features has things scoped *more narrowly*, when was then inconsistent with the rest of the language. 

Given that things were less consistent *and* less useful, we changed course there. 

We may choose to be less consistent if the value is there. But if it's not there, or is marginal, we prefer consistent.

**Reply by CyrusNajmabadi** [reply]
> Readonly by default for non-record class PC parameters fits that strategy. 

That strategy refers to being able to produce values.  In the domain of values, we do agree that being readonly/immutable is preferable.  

However, primary constructors is very much about a feature that is being added for the vast type space that is entirely *not* about values.  And, as such, making it behave like records is something we ultimately decided against.

**Reply by CyrusNajmabadi** [reply]
> there has been an ideology shift within the language team away from favouring immutability. This is really disappointing.

Fyi that's there's been no ideological shift. You're confusing the position we took specifically around a sub domain of the language and consumption space and broadly expanding that out to the entirety of the language. 

We believe there are times where readonly-ness by default makes sense.  And in those spaces, we wanted the terse and elegant syntax to align with that. But that very much does *not* expand to areas that already exist and are already used to dealing with mutability by default.

**Reply by HaloFour** [reply]
> scoping to the current block is consistent with the rest of the language. 

I have to disagree with this statement.  Previous to `out var` and pattern variables, control statements that would introduce new variables into scope would only do so for the following embedded statement.  That includes `for`, `foreach`, `using` and `catch`.  I understand _why_ the team made the decision that they did, but it did create an inconsistency in the language, to the extent that now the a single statement can now introduce multiple variables with completely different overlapping scopes.  Narrow scoping would definitely have been more consistent.  I much prefer the route Java took here, with "flow scoping".

**Reply by CyrusNajmabadi** [reply]
I would agree, of it was the control construct introducing the variable (similar to the above).  For example, if it was:

`if (var v = GetExpensiveValue(); v != null)`

Where there were special control-variable forms. But it's not. It's just an out param and where the local for it should go (similar to the location for an equivalent local-decl or not). 

This raises a common question of: "what to be consistent with" though, and different people latch into different parts of the language. 

Note: I think there is support for doing something similar to java here. It's just not risen to the level of being worth investing in. That follows our general strategy if waiting to see if community really needs something or if the simple approach is sufficient. So far, the latter has held.

**Reply by HaloFour** [reply]
This is a pointless tangent, but that's what makes it so much fun.  😁 

> Where there were special control-variable forms. But it's not. It's just an out param and where the local for it should go (similar to the location for an equivalent local-decl or not).

IMO, I don't think that matters, any more than it does for variables declared within a `for` block (which are completely optional).  The variables in these cases are still declared in the same place, so having some of those variables declared for the narrow scope and some for the outer scope is inconsistent and jarring.  Sure, you can make up new rules for a completely new language feature, but that is not keeping consistent with the rest of the language, it's doing something completely new.

> Note: I think there is support for doing something similar to java here.

I guess it'd have to be called "anti-flow shadowing"?  😄

**Reply by CyrusNajmabadi** [reply]
> IMO, I don't think that matters,

I think it does.  Take, for example, the `for` construct.  It *already* has a plce to put variables.  So, if you use an `out-var`it goes into that scope.  So, for the sake of example, if you had this very weird code:

```c#

        for(bool b = true, c = int.TryParse("1", out int result1); int.TryParse("2", out var result2); )
        {
            Console.WriteLine(result1);
            Console.WriteLine(result2);
        }

        // Not legal.
        Console.WriteLine(result1);
        Console.WriteLine(result2);
```

You'd see that `result1/result2` are scoped *inside* teh `for` (like the existing variables it can declare).

Basically, we were consistent with scoping new variables in the location they would have been scoped in *before*.  You can argue that we were inconsistent by *not* introducing new scopes for these embedded-statement-owners to match some of the other flow-constructs.  Both are reasonable forms of consistency to argue for.  We ended up going with the former because, well, it made much more sense to us to fit with existing scoping and to make it so that we would then not have to then go add *more* constructs to the langauge whose only purpose was to do the same, but with wider scope.

> I guess it'd have to be called "anti-flow shadowing"? 

Sure, why not :D

**Reply by HaloFour** [reply]
> Basically, we were consistent with scoping new variables in the location they would have been scoped in _before_.

As if you declared the variable _outside_ of the control statement.  But you're not declaring the variable _outside_ of the control statement.  Pretending it's like that only exposes the warts of the implementation, and still creates scenarios where variables declared literally side-by-side within the same construct have completely different scopes.

The argument as to where the variable is actually declared was the same one made by C compilers that held onto wider scoping for `for` loops.  ANSI settled that, and C# copied the ANSI behavior.

**Reply by CyrusNajmabadi** [reply]
> As if you declared the variable outside of the control statement.

Right.  We've been consistent with declaring the variable in the innermost scope where you would have had to declare it yourself.  

> But you're not declaring the variable outside of the control statement. 

?   

Maybe we're talking past each other, but (unless you're talking about for/foreach) we very much are.  Note that for constructs that *do* have a scope to declare things in, we *do* declare in that scope.  `foreach` works thsi was as well:

```c#
        foreach (var x in GetEnumerator(out var other))
        {
        }

        // Not legal either
        Console.WriteLine(other);
```

> Pretending it's like that only exposes the warts of the implementation

I'm not sure what you mean by 'pretend'.  There is no pretending here.  We decided to be consistent with the scopes that existed prior to out-vars.  `for/foreach/using` all have scopes to place their variables into.  `out-vars` properly respect that and go into those scopes.

`if` simply has no variable scopes, period.  So anything declared within its condition just goes up to the nearest scope and places into that.  

--

I do get the argument that we're inconsistent in that not *all* flow constructs declare scopes.  That would have been another way to get consistency.  But choosing to not introduce new scopes and just use the ones that exist today is consistent with how scopes have always worked.

**Reply by HaloFour** [reply]
@CyrusNajmabadi 

> We decided to be consistent with the scopes that existed prior to out-vars.

Oops, I forgot that decision was made.  Shows how often I use patterns or `out` vars in loops.  😅 

But I do think that it's weird that `if` is treated differently just because there was no syntax that allowed to declare a variable in that control statement before.  Feels like it should have the same scoping rules as all of the other control statements.

Anyway, tangent, obviously not being reconsidered at this point.

**Reply by CyrusNajmabadi** [reply]
> Oops, I forgot that decision was made. Shows how often I use patterns or out vars in loops. 

Right.  It was a decision to be consistent with the past.  It was also very easy to explain, and it also made several core scenarios just work out nicely without any extra effort. :)

> Feels like it should have the same scoping rules as all of the other control statements.

Yes.  That was another 'consistency' direction we could have gone with (and you know we cetainly did consider that and even prototype that :)).

**Comment by RenderMichael** [comment]
I respect the LDM's decision on `readonly`-ness of primary constructors, but I will have to abstain from using them in C# 12. Hopefully the open questions surrounding it can be hashed out By C# 13 and my team can use this feature outside of preview by late 2024.
